src/data_chain/block.rs

```diff
@@ -189,90 +188,129 @@ mod tests {
     use maidsafe_utilities::SeededRng;
     use rand::Rng;
     use rust_sodium;
+    use std::collections::BTreeMap;
 
     #[test]
     fn create_then_add_proofs() {
         let mut rng = SeededRng::thread_rng();
         unwrap!(rust_sodium::init_with_rng(&mut rng));
 
-        let mut full_info_0 = FullInfo::node_new(0);
-        let mut full_info_1 = FullInfo::node_new(0);
-        let public_info_0 = *full_info_0.public_info();
-        let public_info_1 = *full_info_1.public_info();
+        let full_info_0 = FullInfo::node_new(1);
+        let full_info_1 = FullInfo::node_new(rng.gen_range(0, 255));
 
         let payload = "Gone";
         let vote0 = unwrap!(Vote::new(full_info_0.secret_sign_key(), payload));
-        assert!(vote0.validate_signature(&public_info_0));
         let vote1 = unwrap!(Vote::new(full_info_1.secret_sign_key(), payload));
-        assert!(vote1.validate_signature(&public_info_1));
-
-        full_info_0.set_age(1u8);
-        full_info_1.set_age(rng.gen_range(0, 255));
-        let public_info_00 = *full_info_0.public_info();
-        let public_info_10 = *full_info_1.public_info();
 
-        let proof0 = unwrap!(vote0.proof(&public_info_00));
+        let proof0 = unwrap!(vote0.proof(full_info_0.public_info()));
         assert!(proof0.validate_signature(&payload));
-        let proof1 = unwrap!(vote1.proof(&public_info_10));
+        let proof1 = unwrap!(vote1.proof(full_info_1.public_info()));
         assert!(proof1.validate_signature(&payload));
-        let mut b0 = unwrap!(Block::new(&vote0, &public_info_00));
-        assert!(proof0.validate_signature(&b0.payload));
-        assert!(proof1.validate_signature(&b0.payload));
+
+        let mut block = unwrap!(Block::new(&vote0, full_info_0.public_info()));
+        assert_eq!(*block.payload(), payload);
+        assert_eq!(block.num_proofs(), 1);
+
         let mut valid_voters = BTreeSet::new();
-        let _ = valid_voters.insert(public_info_00);
-        let _ = valid_voters.insert(public_info_10);
-        assert!(b0.add_proof(proof0, &valid_voters).is_err());
-        assert_eq!(b0.num_proofs(), 1);
-        assert!(b0.add_proof(proof1, &valid_voters).is_ok());
-        assert_eq!(b0.num_proofs(), 2);
+        let _ = valid_voters.insert(*full_info_0.public_info());
+        let _ = valid_voters.insert(*full_info_1.public_info());
+
+        assert!(block.add_proof(proof0, &valid_voters).is_err());
```

Comment:
Maybe it'd make sense to make the asserted error more explicit? I.e. something like this:
```
match block.add_proof(proof0, &valid_voters) {
  RoutingError::DuplicateSignatures => {},
  e => panic!("Unexpected result {}", e),
}
```

And also test other error conditions.
(This is more of a comment for further consideration/refactoring though, feel free to ignore)

Reply:
:+1: